### ISSUE TITLE
 How do you train a model without softmax?

### ISSUE THREAD

**Issue by dgl-prc** [eth-sri/eran#25]
I try to verify a model that is trained with softmax, but the ERAN does not support the softmax operation. I inspected the models provided by the ERAN officially and find that they all do not include the softmax operation. 

So I would like to know how did you train a model for classification without softmax? Did you use a loss function designed by yourself?  Here is my code:

```import tensorflow as tf
import numpy as np
import pandas as pd


def dense_to_one_hot(labels_dense, num_classes):
    num_labels = labels_dense.shape[0]
    index_offset = np.arange(num_labels) * num_classes
    labels_one_hot = np.zeros((num_labels, num_classes))
    labels_one_hot.flat[index_offset + labels_dense.ravel()] = 1
    return labels_one_hot


def load_data(data_path="../../data/mnist/mnist_train.csv"):
    data = pd.read_csv(data_path)
    images = data.iloc[:, 1:].values
    images = images.astype(np.float)
    images = np.multiply(images, 1.0 / 255.0)

    labels_flat = data[['label']].values.ravel()
    num_classes = np.unique(labels_flat).shape[0]

    labels = dense_to_one_hot(labels_flat, num_classes)

    return images, labels


def create_model():
    x = tf.placeholder("float", shape=[1, 784])
    y = tf.placeholder("float", shape=[1, 10])  # maybe is [1,10]

    l1 = tf.Variable(tf.zeros([784, 10]))
    b1 = tf.Variable(tf.zeros([10]))
    output = tf.add(tf.matmul(x, l1), b1)

    # l2 = tf.Variable(tf.zeros([50, 50]))
    # b2 = tf.Variable(tf.zeros([50]))
    # a2 = tf.nn.relu(tf.add(tf.matmul(a1, l2), b2))

    # l3 = tf.Variable(tf.zeros([50, 10]))
    # b3 = tf.Variable(tf.zeros([10]))
    # output = tf.matmul(a1, l3)+b3

    # W = tf.Variable(tf.zeros([784, 10]))
    # b = tf.Variable(tf.zeros([10]))
    # output = tf.matmul(x, W) + b

    logits = tf.nn.softmax(output)
    loss = -tf.reduce_sum(y * tf.log(logits))

    return loss, x, y,output,logits


def next_batch(images, labels, batch_size):
    for i in range(0, len(labels), batch_size):
        batch_x = images[i:i + batch_size]
        batch_y = labels[i:i + batch_size]
        yield batch_x, batch_y


def train():
    images, labels = load_data("../../data/mnist/mnist_train.csv")
    test_images, test_labels = load_data("../../data/mnist/mnist_train.csv")
    obj_fun, x, y, output, logits = create_model()
    sess = tf.InteractiveSession()
    sess.run(tf.initialize_all_variables())
    train_step = tf.train.GradientDescentOptimizer(0.01).minimize(obj_fun)
    for batch_x, batch_y in next_batch(images, labels, batch_size=1):
        sess.run([train_step, obj_fun], feed_dict={x: batch_x, y: batch_y})

    # eval
    predict = tf.nn.relu(logits)
    saver = tf.train.Saver()
    saver.save(sess, "./trained_model/mymodel")


if __name__ == '__main__':
    train()
```
Note that the code before saving, i.e. predict = tf.nn.relu(logits), is not used actually. I add this code because the ERAN model only can translate a graph which is ended with a qualified operation. If I don't add this, the ERAN will raise an exception when executes "tf.import_graph_def(new_graph_def)" at line 135 in tf_verify/tensorflow_translator.py

**Comment by LebronX** [comment]
> I try to verify a model that is trained with softmax, but the ERAN does not support the softmax operation. I inspected the models provided by the ERAN officially and find that they all do not include the softmax operation.
> 
> So I would like to know how did you train a model for classification without softmax? Did you use a loss function designed by yourself? Here is my code:
> 
> ```
> import numpy as np
> import pandas as pd
> 
> 
> def dense_to_one_hot(labels_dense, num_classes):
>     num_labels = labels_dense.shape[0]
>     index_offset = np.arange(num_labels) * num_classes
>     labels_one_hot = np.zeros((num_labels, num_classes))
>     labels_one_hot.flat[index_offset + labels_dense.ravel()] = 1
>     return labels_one_hot
> 
> 
> def load_data(data_path="../../data/mnist/mnist_train.csv"):
>     data = pd.read_csv(data_path)
>     images = data.iloc[:, 1:].values
>     images = images.astype(np.float)
>     images = np.multiply(images, 1.0 / 255.0)
> 
>     labels_flat = data[['label']].values.ravel()
>     num_classes = np.unique(labels_flat).shape[0]
> 
>     labels = dense_to_one_hot(labels_flat, num_classes)
> 
>     return images, labels
> 
> 
> def create_model():
>     x = tf.placeholder("float", shape=[1, 784])
>     y = tf.placeholder("float", shape=[1, 10])  # maybe is [1,10]
> 
>     l1 = tf.Variable(tf.zeros([784, 10]))
>     b1 = tf.Variable(tf.zeros([10]))
>     output = tf.add(tf.matmul(x, l1), b1)
> 
>     # l2 = tf.Variable(tf.zeros([50, 50]))
>     # b2 = tf.Variable(tf.zeros([50]))
>     # a2 = tf.nn.relu(tf.add(tf.matmul(a1, l2), b2))
> 
>     # l3 = tf.Variable(tf.zeros([50, 10]))
>     # b3 = tf.Variable(tf.zeros([10]))
>     # output = tf.matmul(a1, l3)+b3
> 
>     # W = tf.Variable(tf.zeros([784, 10]))
>     # b = tf.Variable(tf.zeros([10]))
>     # output = tf.matmul(x, W) + b
> 
>     logits = tf.nn.softmax(output)
>     loss = -tf.reduce_sum(y * tf.log(logits))
> 
>     return loss, x, y,output,logits
> 
> 
> def next_batch(images, labels, batch_size):
>     for i in range(0, len(labels), batch_size):
>         batch_x = images[i:i + batch_size]
>         batch_y = labels[i:i + batch_size]
>         yield batch_x, batch_y
> 
> 
> def train():
>     images, labels = load_data("../../data/mnist/mnist_train.csv")
>     test_images, test_labels = load_data("../../data/mnist/mnist_train.csv")
>     obj_fun, x, y, output, logits = create_model()
>     sess = tf.InteractiveSession()
>     sess.run(tf.initialize_all_variables())
>     train_step = tf.train.GradientDescentOptimizer(0.01).minimize(obj_fun)
>     for batch_x, batch_y in next_batch(images, labels, batch_size=1):
>         sess.run([train_step, obj_fun], feed_dict={x: batch_x, y: batch_y})
> 
>     # eval
>     predict = tf.nn.relu(logits)
>     saver = tf.train.Saver()
>     saver.save(sess, "./trained_model/mymodel")
> 
> 
> if __name__ == '__main__':
>     train()
> ```
> 
> Note that the code before saving, i.e. predict = tf.nn.relu(logits), is not used actually. I add this code because the ERAN model only can translate a graph which is ended with a qualified operation. If I don't add this, the ERAN will raise an exception when executes "tf.import_graph_def(new_graph_def)" at line 135 in tf_verify/tensorflow_translator.py

Hi,

Did you solve this question? I have the same doubt.

Thanks

**Comment by dgl-prc** [comment]
I solved it by moving the softmax out of the definition of the model

**Comment by LebronX** [comment]
Thanks!
